aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,129 @@
+ * @param char_offset the position to search for in the buffer list, describing the zero referenced
+ *      character index if all buffer strings were concatenated end to end
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer.
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+            size_t char_offset, size_t *entry_offset_byte_rtn )
+{
+    size_t bytes_traversed = 0;
+    uint8_t traverse_len = 0;
+
+    if (buffer != NULL)
+    {
+    	int pos = buffer->out_offs;
+    	if (buffer->full)
+    	{
+    		traverse_len = AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+    	}
+    	else
+    	{
+    		if (buffer->in_offs > buffer->out_offs)
+    		{
+    			traverse_len = buffer->in_offs - buffer->out_offs;
+    		}
+    		else
+    		{
+    			traverse_len = buffer->out_offs - buffer->in_offs;
+    		}
+    		
+    	}
+	while (traverse_len > 0)
+	{
+    		if ((char_offset >= 0) && ((bytes_traversed + buffer->entry[pos].size) > char_offset))
+    		{
+    			*entry_offset_byte_rtn = (char_offset - bytes_traversed);
+    			return &buffer->entry[pos];
+    		}
+    		
+    		bytes_traversed += buffer->entry[pos].size;
+
+    		if (pos == AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED-1)
+    		{
+    			pos = 0;
+    		}
+    		else
+    		{
+    			pos++;	
+    		}
+    		traverse_len--;
+	}
+    }
+    
+    return NULL;
+}
+
+/**
+* Adds entry @param add_entry to @param buffer in the location specified in buffer->in_offs.
+* If the buffer was already full, overwrites the oldest entry and advances buffer->out_offs to the
+* new start location.
+* Any necessary locking must be handled by the caller
+* Any memory referenced in @param add_entry must be allocated by and/or must have a lifetime managed by the caller.
+*/
+void aesd_circular_buffer_add_entry(struct aesd_circular_buffer *buffer, const struct aesd_buffer_entry *add_entry)
+{
+    if ((buffer == NULL) || (add_entry == NULL))
+    {
+    	return;
+    }
+
+    buffer->entry[buffer->in_offs] = *add_entry;
+    if (buffer->full)
+    {
+    	if (buffer->in_offs == AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED -1 )
+    	{
+    		buffer->in_offs=0;
+    		buffer->out_offs=0;
+    	}
+    	else
+    	{
+    		buffer->in_offs += 1;
+    		buffer->out_offs = buffer->in_offs;
+    	}
+    }
+    else
+    {
+    	if (buffer->in_offs == AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED - 1)
+    	{
+    		buffer->in_offs=0;
+    		buffer->full = true;
```

Comment:
@charan1811 I don't think this is the correct logic - what does this assume about the location of `out_offs`?

Reply:
Out_offs should be checked for zero here